chore(lambda): throw a relevant error message when asset is used across stacks

nija-at opened the pull request:
Although the constructor of the AssetCode doesn't imply it's real
dependencies, an AssetCode currently can only be associated to consumers
within a single stack.

Attempting to use them across stacks results in the conspicuous error -
"Cannot reference across apps. Consuming and producing stacks must be
defined within the same CDK app."

The correct solution here would have been to model this such that the
stack dependency is made explicit on the customer facing API, perhaps by
taking a 'scope' parameter.

For the moment, throw a more relevant error message so that it's obvious to
the user on what is wrong.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7db44e9975e246ffe045961b95b5d78cf7537136
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2936b864-f914-4e08-a539-06d43ae65acd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/@aws-cdk/aws-lambda/lib/code.ts:
```diff
@@ -242,35 +242,43 @@ export class AssetCode extends Code {
-    // If the same AssetCode is used multiple times, retain only the first instantiation.
-    if (!this.asset) {
-      this.asset = new s3_assets.Asset(scope, 'Code', {
+    const stackPath = cdk.Stack.of(scope).node.path; // use path instead of name to account for Stage.
```
Is this good enough? Wouldn't the same stack path in multiple `Apps` still give the same error?

rix0rrr replied:
Aha well. This change DOES address the issue it's purporting to address (use same code in multiple stacks inside the same app), but it does NOT address the error you mentioned, and would still fail if you reused a single `Code` object among multiple unit tests.

```
Cannot reference across apps. Consuming and producing stacks must be defined within the same CDK app.
```

rix0rrr commented on packages/@aws-cdk/aws-lambda/lib/code.ts:
```diff
@@ -232,7 +232,7 @@ export class InlineCode extends Code {
 export class AssetCode extends Code {
   public readonly isInline = false;
-  private asset?: s3_assets.Asset;
+  private assets: Record<string, s3_assets.Asset> = {};
```
Why not make this a `Map<Stack, s3_assets.Asset>`?

Simpler and more direct.

nija-at commented:
I've changed my mind on this. Something feels wrong in doing this cache and allowing an asset to be used anywhere (any `App`, any `Stack`). It feels like it might come back to bite me.

Instead, I've updated the code to throw a more relevant error message instead of the conspicuous one.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 98b1bc13e2196134dddf29faef793d00bbb85d00
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=aeee8494-69ae-4ad4-8272-d263bc62c17b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/@aws-cdk/aws-lambda/lib/code.ts:
```diff
@@ -248,6 +248,8 @@ export class AssetCode extends Code {
         ...this.options,
       });
+    } else if (cdk.Stack.of(this.asset) !== cdk.Stack.of(scope)) {
+      throw new Error(`Asset is already associated with another stack '${cdk.Stack.of(this.asset).stackName}'.`);
```
When writing an error message, try to also suggest what the user needs to do to fix it (even if it's super obvious to you, it might not be to the user seeing the error). Add something like:

"Create a new Code instance for every stack."

Provisional approval but I hope you'll consider this.

eladb reviewed (commented):
LGTM

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 6be0a4e2ecb9e17304a9c46bbb23db0065c335ad
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=23db86ad-3805-401d-ad3e-29a534ff9011%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b4c2de5bd223bfcec009eeb1e6f648fa3c51757c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=dd841995-3b3f-4f6e-b2dd-f36e4b52b961%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
